onnxruntime/core/providers/qnn/builder/qnn_fusions.cc

```diff
@@ -0,0 +1,294 @@
+  const GraphViewer& graph_viewer = qnn_model_wrapper.GetGraphViewer();
+
+  // Looking for a standalone DQ to start the sequence.
+  if (start_node_unit.OpType() != QDQ::DQOpName || start_node_unit.UnitType() != NodeUnit::Type::SingleNode) {
+    return Status::OK();
+  }
+
+  const Node& dq_node = start_node_unit.GetNode();
+
+  // DQ must have a single Q child. DQ must not produce a graph output.
+  auto children = graph_utils::FindChildrenByType(dq_node, QDQ::QOpName);
+  if (children.size() != 1 || dq_node.GetOutputEdgesCount() != 1 || graph_viewer.NodeProducesGraphOutput(dq_node)) {
+    return Status::OK();
+  }
+
+  const Node& q_node = *children[0];
+  const auto q_node_unit_it = node_unit_map.find(&q_node);
+
+  ORT_RETURN_IF(q_node_unit_it == node_unit_map.end(), "Node does not have a corresponding NodeUnit");
+
+  const NodeUnit* q_node_unit = q_node_unit_it->second;
+
+  // Check if Q node has already been handled. Should not be the case if this
+  // fusion function has been called in topological order, but check to be safe.
+  if (handled_node_units.count(q_node_unit) != 0) {
+    return Status::OK();
+  }
+
+  // Q child must not already be part of a QDQ NodeUnit (i.e., be standalone).
+  if (q_node_unit->UnitType() != NodeUnit::Type::SingleNode) {
+    return Status::OK();
+  }
+
+  auto get_const_initializer = [&graph_viewer](const std::string& initializer_name) {
+    return graph_viewer.GetConstantInitializer(initializer_name, true);
+  };
+
+  // DQ and Q must have equal scale type and different zp type.
+  if (!QDQ::IsDQQConversion(dq_node, q_node, get_const_initializer, graph_viewer.ModelPath())) {
+    return Status::OK();
+  }
+
+  const auto& node_name = utils::GetNodeName(start_node_unit);
+  const NodeUnitIODef& input_def = start_node_unit.Inputs()[0];
+  const NodeUnitIODef& output_def = q_node_unit->Outputs()[0];
+
+  QnnTensorWrapper input_tensor;
+  QnnTensorWrapper output_tensor;
+
+  // Run QNN validation on the final fused node before committing to doing a fusion.
+  // Importantly, this validation process does not modify the qnn_model_wrapper.
+  // If validation fails here, we return Status::OK() to allow QNN EP to use the normal OpBuilder workflow.
+  QNN_RETURN_OK_IF_ERROR(qnn_model_wrapper.MakeTensorWrapper(input_def, input_tensor), logger);
+  QNN_RETURN_OK_IF_ERROR(qnn_model_wrapper.MakeTensorWrapper(output_def, output_tensor), logger);
+  QNN_RETURN_OK_IF_ERROR(qnn_model_wrapper.ValidateQnnNode(node_name,
+                                                           QNN_OP_PACKAGE_NAME_QTI_AISW,
+                                                           QNN_OP_CONVERT,
+                                                           {input_tensor.GetQnnTensor()},
+                                                           {output_tensor.GetQnnTensor()},
+                                                           {}),
+                         logger);
```

Comment:
I want to draw attention to the above lines. These fusions call QNN's validator api before deciding to actually do the fusion. This happens even when compiling the graph.

The validation does not modify any state, and if it fails, we return Status::OK() to allow the EP to use the normal OpBuilder workflow.